compiler/objectfmt/OMRFunctionCallData.hpp

```diff
@@ -0,0 +1,118 @@
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_FUNCTIONCALLDATA_INCL
+#define OMR_FUNCTIONCALLDATA_INCL
+
+/*
+ * The following #define and typedef must appear before any #includes in this file
+ */
+#ifndef OMR_FUNCTIONCALLDATA_CONNECTOR
+#define OMR_FUNCTIONCALLDATA_CONNECTOR
+namespace OMR { class FunctionCallData; }
+namespace OMR { typedef OMR::FunctionCallData FunctionCallDataConnector; }
+#endif
+
+#include <stddef.h>
+#include <stdint.h>
+#include "infra/Annotations.hpp"
+
+namespace TR { class CodeGenerator; }
+namespace TR { class Node; }
+namespace TR { class SymbolReference; }
+
+namespace OMR
+{
+
+/**
+ * @class FunctionCallData
+ *
+ * @details
+ *    The \c FunctionCallData class is an extensible class used to encapsulate
+ *    all the information at particular call site needed to emit a function
+ *    call in a particular object format.  Each platform and language
+ *    environment is responsible for extending this class to capture whatever
+ *    meaningful data is necessary at a call site.
+ *
+ *    The resulting \c TR::FunctionCallData class is accepted as a parameter to
+ *    most \c TR::ObjectFormat APIs.
+ */
+class OMR_EXTENSIBLE FunctionCallData
+   {
+
+public:
+
+   FunctionCallData(
+         TR::SymbolReference *parm_methodSymRef,
+         TR::Node *parm_callNode,
+         TR::CodeGenerator *parm_cg) :
+      methodSymRef(parm_methodSymRef),
+      callNode(parm_callNode),
+      bufferAddress(0),
+      out_encodedMethodAddressLocation(0),
+      cg(parm_cg) {}
+
+   FunctionCallData(
+         TR::SymbolReference *parm_methodSymRef,
+         TR::Node *parm_callNode,
+         uint8_t *parm_bufferAddress,
+         TR::CodeGenerator *parm_cg) :
+      methodSymRef(parm_methodSymRef),
+      callNode(parm_callNode),
+      bufferAddress(parm_bufferAddress),
+      out_encodedMethodAddressLocation(0),
+      cg(parm_cg) {}
+
+   /**
+    * The \c TR::SymbolReference of the method to call
+    */
+   TR::SymbolReference *methodSymRef;
+
+   /**
+    * If applicable, the \c TR::Node of the IL opcode that requires this call
+    *
+    * If not applicable, the field is NULL.
+    */
+   TR::Node *callNode;
+
+   /**
+    * The \c TR::CodeGenerator object
+    */
+   TR::CodeGenerator *cg;
+
+   /**
+    * The address of a buffer where the function call is to be emitted,
+    * or 0 if the buffer address is unknown or not required.
+    */
+   uint8_t *bufferAddress;
+
+   /**
+    * If the address of the funtion is known and it is encoded somewhere
+    * as part of the function dispatch sequence then this field records
+    * the address where that function address is written.
+    */
+   uint8_t *out_encodedMethodAddressLocation;
+
```

Comment:
We should likely make this a `struct` then? Minor detail I suppose. `struct` fields are public by default where as `class` fields are private by default.

Reply:
Yes, you're right.  But I left it as a class because it is decorated `OMR_EXTENSIBLE`.  I didn't think through all the implications of making "extensible structs".  In theory we shouldn't have a problem (maybe some work on the linter?) but perhaps something to be solved in a different PR.